### PR TITLE
Root Table

### PR DESCRIPTION
**Pull request by sundarshankar89**
This PR Introduces `TableGraph` which exposes public methods

- `getRootTables`
- `getUpstreamTables`
- `getDownstreamTables`

Closes https://github.com/databrickslabs/remorph/issues/23

### CODE REVIEW

File: core/src/main/scala/com/databricks/labs/remorph/graph/lineage.scala [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package com.databricks.labs.remorph.graph
+
+import com.databricks.labs.remorph.discovery.TableDefinition
+
+trait DependencyGraph {
+  protected def addNode(id: TableDefinition, metadata: Map[String, String]): Unit
+  protected def addEdge(from: TableDefinition, to: TableDefinition, metadata: Map[String, String]): Unit
+  def getUpstreamTables(table: String): Set[TableDefinition]
+  def getDownstreamTables(table: String): Set[TableDefinition]
+  def getRoot(table: String, level: Int): TableDefinition
```
Comment (nfx) [inline core/src/main/scala/com/databricks/labs/remorph/graph/lineage.scala:10]:
```suggestion
```

File: core/src/main/scala/com/databricks/labs/remorph/graph/lineage.scala [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package com.databricks.labs.remorph.graph
+
+import com.databricks.labs.remorph.discovery.TableDefinition
+
+trait DependencyGraph {
+  protected def addNode(id: TableDefinition, metadata: Map[String, String]): Unit
+  protected def addEdge(from: TableDefinition, to: TableDefinition, metadata: Map[String, String]): Unit
+  def getUpstreamTables(table: String): Set[TableDefinition]
+  def getDownstreamTables(table: String): Set[TableDefinition]
```
Comment (nfx) [inline core/src/main/scala/com/databricks/labs/remorph/graph/lineage.scala:9]:
```suggestion
  def getUpstreamTables(table: TableDefinition): Set[TableDefinition]
  def getDownstreamTables(table: TableDefinition): Set[TableDefinition]
```

File: core/src/main/scala/com/databricks/labs/remorph/graph/TableDependencyGraph.scala [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+      plan match {
+        case Result.Success(p) => buildNode(p, tableDefinition, query)
+          generateEdges(p, tableDefinition, query.id)
+        case _ => logger.warn(s"Failed to produce plan from query: ${query.source}")
+      }
+    }
+  }
+
+  // TODO Implement logic for fetching edges(parents) only upto certain level
+  override def getRoot(table: String, level: Int = 1): TableDefinition = {
```
Comment (nfx) [inline core/src/main/scala/com/databricks/labs/remorph/graph/TableDependencyGraph.scala:119]: there are multiple roots. the simplest logic should be:
- count number of indegrees for a node. indegree is an incoming write to a table.
- tables with zero indegrees is a root, as nothing writes to it.

File: core/src/main/scala/com/databricks/labs/remorph/graph/TableDependencyGraph.scala [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+import com.databricks.labs.remorph.discovery.{ExecutedQuery, QueryHistory, TableDefinition}
+import com.databricks.labs.remorph.parsers.PlanParser
+import com.databricks.labs.remorph.transpilers.{Result, SourceCode}
+import com.typesafe.scalalogging.LazyLogging
+import com.databricks.labs.remorph.parsers.{intermediate => ir}
+
+protected case class Node(tableDefinition: TableDefinition, metadata: Map[String, Set[String]])
+protected case class Edge(from: TableDefinition, to: TableDefinition, metadata: Map[String, String])
+
+class TableDependencyGraph(parser: PlanParser[_]) extends DependencyGraph with LazyLogging {
```
Comment (nfx) [inline core/src/main/scala/com/databricks/labs/remorph/graph/TableDependencyGraph.scala:12]:
```suggestion
class TableGraph(parser: PlanParser[_], tables: Set[TableDefinition]) extends DependencyGraph with LazyLogging {
```
shorter names are always better. also, i think it's better to depend on a precomputed set of table definitions, that we retrieve separately.

Comment (sundarshankar89) [reply core/src/main/scala/com/databricks/labs/remorph/graph/TableDependencyGraph.scala:12]: isn't it better to pass the `tabledefinition` to the public method for this class. That way we create instance of one object per sourcesystem.

File: core/src/main/scala/com/databricks/labs/remorph/graph/lineage.scala (hunk 2 above)
Comment (sundarshankar89) [reply core/src/main/scala/com/databricks/labs/remorph/graph/lineage.scala]: Done

File: core/src/main/scala/com/databricks/labs/remorph/graph/TableGraph.scala [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+          .toSet
+      case None =>
+        logger.warn(s"Table ${table.table} not found in the graph")
+        Set.empty[TableDefinition]
+    }
+  }
+
+  // TODO implement recursive call to get all downstream tables.
+  override def getDownstreamTables(table: TableDefinition): Set[TableDefinition] = {
+    val node = nodes.find(_.tableDefinition == table)
```
Comment (sundarshankar89) [inline core/src/main/scala/com/databricks/labs/remorph/graph/TableGraph.scala:175]: ignore

File: core/src/main/scala/com/databricks/labs/remorph/graph/TableDependencyGraph.scala (hunk 3 above)
Comment (sundarshankar89) [reply core/src/main/scala/com/databricks/labs/remorph/graph/TableDependencyGraph.scala]: Done

File: core/src/main/scala/com/databricks/labs/remorph/graph/TableGraph.scala [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+protected case class Node(tableDefinition: TableDefinition, metadata: Map[String, Set[String]])
+// `from` is the table which is sourced to create `to` table
+protected case class Edge(from: Node, to: Option[Node], metadata: Map[String, String])
+
+class TableGraph(parser: PlanParser[_]) extends DependencyGraph with LazyLogging {
+  private val nodes = scala.collection.mutable.Set.empty[Node]
+  private val edges = scala.collection.mutable.Set.empty[Edge]
+
+  override protected def addNode(id: TableDefinition, metadata: Map[String, Set[String]]): Unit = {
+    // Metadata list of query ids and add node only if it is not already present.
```
Comment (nfx) [inline core/src/main/scala/com/databricks/labs/remorph/graph/TableGraph.scala:18]: can we avoid free-form string-to-string maps and add explicit `queryIds: Set[String]`?

**Review by nfx — APPROVED** [review]
lgtm

**Pull request merged**